lib/Parser.js

```diff
@@ -37,6 +38,107 @@ const POSSIBLE_AST_OPTIONS = [{
 	}
 }];
 
+class StackedSetMap {
+	constructor(defaultValue, parentStack) {
+		this.defaultValue = defaultValue;
+		this.stack = parentStack === undefined ? [] : parentStack.slice();
+		this.map = new Map();
+		this.stack.push(this.map);
+	}
+
+	add(item) {
+		this.map.set(item, true);
+	}
+
+	set(item, value) {
+		this.map.set(item, value);
+	}
+
+	delete(item) {
+		this.map.set(item, false);
```

Comment:
This collection maybe could be made more generic by using two marker objects (2 module scope objects that are not used for anything but their `===` identity).

Or maybe Symbol could be appropriate. ATM setting any falsy value in the collection is effectively the same as deleting it, which does happen to work in the current use case.